jwst/picture_frame/picture_frame.py

```diff
@@ -0,0 +1,230 @@
+UPPER_EDGE_REGION = (1985, 2043)
+CENTER_REGION = (76, 1984)
+
+
+def _region_masks(background_mask):
+    """
+    Make region masks for center and edge data.
+
+    Parameters
+    ----------
+    background_mask : ndarray
+        Background mask image (1 = valid background pixel, 0 otherwise).
+
+    Returns
+    -------
+    center_data : ndarray of bool
+        Array matching background mask shape containing True for background pixels
+        in the center region.
+    edge_data : ndarray of bool
+        Array matching background mask shape containing True for background pixels
+        in the edge region.
+    """
+    y, x = np.mgrid[: background_mask.shape[0], : background_mask.shape[1]]
+
+    # Background data in the center region
+    center_data = (
+        background_mask
+        & (y >= CENTER_REGION[0])
+        & (y < CENTER_REGION[1])
+        & (x >= CENTER_REGION[0])
+        & (x < CENTER_REGION[1])
+    )
+
+    # Background data in the edge regions, not including the reference pixels around the border
+    out_of_range = (
+        (y < LOWER_EDGE_REGION[0])
+        | (y >= UPPER_EDGE_REGION[1])
+        | (x < LOWER_EDGE_REGION[0])
+        | (x >= UPPER_EDGE_REGION[1])
+    )
+    bottom = (y >= LOWER_EDGE_REGION[0]) & (y < LOWER_EDGE_REGION[1])
+    top = (y >= UPPER_EDGE_REGION[0]) & (y < UPPER_EDGE_REGION[1])
+    left = (x >= LOWER_EDGE_REGION[0]) & (x < LOWER_EDGE_REGION[1])
+    right = (x >= UPPER_EDGE_REGION[0]) & (x < UPPER_EDGE_REGION[1])
+    edge_data = background_mask & (bottom | top | left | right) & ~out_of_range
+
+    return center_data, edge_data
+
+
+def correct_picture_frame(
+    input_model,
+    pictureframe_model,
+    mask_science_regions=True,
+    n_sigma=2.0,
+    input_dir="",
+    save_mask=False,
+    save_correction=False,
+):
+    """
+    Correct full-frame NIRSpec data for the thermal picture frame effect.
+
+    Parameters
+    ----------
+    input_model : `~stdatamodels.jwst.datamodels.RampModel` \
+                  or `~stdatamodels.jwst.datamodels.ImageModel` \
+                  or `~stdatamodels.jwst.datamodels.CubeModel`
+        Science model to be corrected. Updated in place.
+    pictureframe_model : `~stdatamodels.jwst.datamodels.PictureFrameModel`
+        Picture frame reference model.
+    mask_science_regions : bool, optional
+        Mask regions of the image defined by WCS bounding
+        boxes for slits/slices, as well as any regions known to be
+        affected by failed-open MSA shutters.
+    n_sigma : float, optional
+        N-sigma rejection level for finding outliers.
+    input_dir : str, optional
+        Path to the input directory.  Used by sub-steps (e.g. assign_wcs
+        for NIRSpec MOS data) to find auxiliary data.
+    save_mask : bool, optional
+        Switch to indicate whether the scene mask should be saved.
+    save_correction : bool, optional
+        Switch to indicate whether the scaled correction data should be saved.
```

Comment:
Since this function doesn't actually do the saving, perhaps rename these to return_mask and return_correction?

On a related note, are the mask and correction models memory-intensive?  I guess I don't immediately see the need to propagate these booleans into this function - there's already a conditional for saving these or not at the step level.  So it seems like it'd be equivalent to return these from this function no matter what, and eliminate the need to propagate these flags.